### DISCUSSION TITLE
Different inferred type `t = tuple(); list(t)` and `list(tuple())`

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#7996] [General]
These three functions has three different inferred return types ([playground](https://pyright-play.net/?pyrightVersion=1.1.364&pythonVersion=3.13&strict=true&code=CYUwZgBGD20BQEoBcBYAUBTEAuEC8Ei6WEATiAG4gCGANgPrYCeADiHNgpgMQ4CuLWiADaiALrEs5bH1IA7CLQCWAZ2wcE6dKEgAjaqUSoMWXARmD2mk5nJU6jVu049%2Bl4QFU5AaznQA7nIANBAAdOESNmQgMvKKquqcWmg6EACOfAAeRpK2MbIKymocAkKI1sl2NAzMbHAw0FxYvIgQALQAfPFqwgBylCCkkVUOtez6pE2urZ3d2J4%2BfoHDA6NOcBmZUxAtXLNF8-QAKgACB5FAA)):

```python
def foo():
    t = ()
    reveal_type(t)  # tuple[()]
    return list(t)

def bar():
    t = tuple()
    reveal_type(t)  # tuple[Unknown, ...]
    return list(t)

def qux():
    return list(tuple())
```

```python
reveal_type(foo)    # () -> list[Never]
reveal_type(bar)    # () -> list[Unknown]
reveal_type(qux)    # () -> list[_T@list]
```

`foo`'s return type is understandable: `()`'s type is `tuple[()]` instead of `tuple[Unknown, ...]`/`tuple[Any, ...]` like `tuple()`, since that is [how `tuple.__new__()` is defined](https://github.com/python/typeshed/blob/fc33ba67ba25545e0decb5897b3d56b40c88b0b5/stdlib/builtins.pyi#L933) (`_T_co` fallbacks to `Unknown`/`Any`).

However, only `bar` is reported as having a partially unknown return type. Instead of `Unknown`, the inferred type for `qux` has an unresolved type variable. On the other hand, `qux()` is of type `list[Unknown]`. Shouldn't 	`qux`, too, be reported?

For comparison, Mypy [requires](https://mypy-play.net/?flags=strict&mypy=1.10.0&python=3.12&gist=7f8a9393d11b1b61c88b72e3fbf477c1) that all of the above are annotated explicitly.

**Comment by erictraut** [comment]
With expressions `()` and `tuple()`, we're on pretty firm footing with regard to how type checkers should work according to the typing spec. I think pyright's behavior is correct here.

With `list(tuple())`, we're in much murkier territory. Mypy evaluates this expression as `list[Never]`. Pyright evaluates it as `list[_T@list]`, which isn't what I would expect. If you would like me to look into this in more detail, feel free to file a bug.

For completeness, we should also look at `list(())`. Here, mypy and pyright both evaluate `list[Never]`.

The typing spec doesn't currently provide much guidance for constraint solving, especially in cases where no constraints are provided for a type variable. This is something we should maybe collectively work to spec and standardize in the future.
